plonk/src/verifier.rs

```diff
@@ -0,0 +1,217 @@
+        let l_1_e = self.l1_poly().evaluate(&evaluation_challenge);
+        let p_i_e = self.gate_constraint.pi_x().evaluate(&evaluation_challenge);
+
+
+        #[cfg(test)]
+        println!("Compute r0");
+        let r_0 = p_i_e - l_1_e * alpha * alpha
+            - alpha * (proof.bar_a + proof.bar_ssigma_1 * beta + gamma)
+            * (proof.bar_b + proof.bar_ssigma_2 * beta + gamma) * (proof.bar_c + gamma) * proof.bar_z_w;
+
+        #[cfg(test)]
+        println!("Compute [D]");
+
+        let d_line1 = q_m_c.mul(proof.bar_a * proof.bar_b) + q_l_c.mul(proof.bar_a) + q_r_c.mul(proof.bar_b)
+        + q_o_c.mul(proof.bar_c) + q_c_c;
+
+        let d_line2 = proof.z_commit.mul((proof.bar_a + beta * evaluation_challenge + gamma)
+            * (proof.bar_b + beta * self.copy_constraint.k1() * evaluation_challenge + gamma)
+            * (proof.bar_c + beta * self.copy_constraint.k2() * evaluation_challenge + gamma)
+            * alpha + l_1_e * alpha * alpha + u);
+
+        let d_line3 = ssigma_3_c.mul((proof.bar_a + beta * proof.bar_ssigma_1 + gamma)
+            * (proof.bar_b + beta * proof.bar_ssigma_2 + gamma) * alpha * beta * proof.bar_z_w);
+
+        let d_line4 = (proof.t_lo_commit + proof.t_mid_commit.mul(Self::power(&evaluation_challenge, proof.degree.clone() + 1) )
+            + proof.t_hi_commit.mul(Self::power(&evaluation_challenge, proof.degree.clone() * 2 + 2))).mul(z_h_e);
+
+        let d = d_line1 + d_line2 - d_line3 - d_line4;
+
+        #[cfg(test)]
+        println!("Compute [F]");
+
+        let f = d + proof.a_commit.mul(v) + proof.b_commit.mul(v * v) + proof.c_commit.mul(v * v * v)
+            + ssigma_1_c.mul(v * v * v * v) + ssigma_2_c.mul(v * v * v * v * v);
+
+        #[cfg(test)]
+        println!("Compute [E]");
+        let e = -r_0 + v * proof.bar_a + v * v * proof.bar_b + v * v * v * proof.bar_c
+            + v * v * v * v * proof.bar_ssigma_1 + v * v * v * v * v * proof.bar_ssigma_2 + u * proof.bar_z_w;
+        let e = scheme.commit_para(&e);
+
+        #[cfg(test)]
+        println!("Compute left side of paring");
+
+        let pairing_left_side = Bls12_381::pairing(
+            (proof.w_ev_x_commit + proof.w_ev_wx_commit.mul(u)).0,
+            self.srs.get_g2s_ref().clone()
+        );
+
+        #[cfg(test)]
+        println!("Compute right side of paring");
+        let pairing_right_side = Bls12_381::pairing(
+            (proof.w_ev_x_commit.mul(evaluation_challenge) + proof.w_ev_wx_commit.mul(u * evaluation_challenge * w)
+                + f - e).0,
+            self.srs.get_g2_ref().clone()
+        );
+
+        #[cfg(test)]
+        println!("Check pairing");
+
+        assert_eq!(pairing_left_side,pairing_right_side, "Verify: Pairing failed, rejected");
+
+        println!("Accepted!!!");
+
+        return true;
+    }
+
+    fn get_circuit_commitment(&self) -> (KzgCommitment, KzgCommitment, KzgCommitment, KzgCommitment,

```

Comment:
How can the verifier get the commitment themself. The right way in my opinion is the verifier get the commitment from the prover